### PR TITLE
Improvement: Allow users to add additional fetch params to token request

### PR DESCRIPTION
**Pull request by bennovakovic**
PR for https://github.com/bpedroza/js-pkce/issues/35

### CODE REVIEW

**Comment by bpedroza** [comment]
After looking at this a bit more, I think I would prefer to change the new parameter to `additionalHeaders` and not allow modification of all the fetch options. Additionally, the default headers should always be preserved and not be allowed to be overridden so we can enforce conformity with the specification.

If you would like to make these change, please go ahead and update this PR, and include some tests for:

1. The new functionality
2. Trying to override default headers without success

Thanks!

**Comment by bennovakovic** [comment]
The `credentials` param is not a header though, so renaming it to `additionalHeaders` doesn't seem right to me. I agree that perhaps limiting the fetch options the user can modify might be a good middle ground.

I am unsure how to proceed based on your feedback, can you please clearly articulate what the expected behaviour should be? 

TL;DR, I want to be able to specify a custom `credentials` options in the fetch call, and this _should not_ be done via headers, but rather as an option to the `fetch` call itself.

**Comment by bpedroza** [comment]
I was confused by the naming of the issue, but I see now after reviewing your example there. 

I am not totally clear why the `credentials` option needs to be changed here. Why is setting the `mode` to `cors` not enough?

The main issue I have with your implementation is that the developer needs to know about this library's dependency, and would need to consult fetch documentation to get CORS configured properly. This adds additional burden to us: changing the dependency would also be a breaking change.

One Idea I have is to create a public method  `allowCORS()` that would set some default configuration for CORS that is merged into all the POST requests' fetch options. I think that would solve the usability and dependency issues I have as well as implement a restriction on which of the fetch options can be altered.

**Comment by bennovakovic** [comment]
I'll give it a test and see if I can get it working with `Access-Control-Allow-Credentials: true` rather than the `credentials` option in `fetch`

**Comment by bennovakovic** [comment]
just setting `mode` to `cors` is not sufficient. I am trying to do a `cors` request _with_ cross domain cookies. The server has been setup to allow credentials via the `Access-Control-Allow-Credentials: true` header, as well as specifcying a correct origin server.

I'm not opposed to a `allowCors()` method, but setting `credentials` to `include` for every request is likely a security flaw. Allowing the client to specify whether they want to `include` them vs `omit` seems a lot more safe to me.

If its too hard, no worries, I will just use my fork of the repo with the changes. I was just trying to be helpful to others who might be trying to do cors + cookies in a secure fashion.

**Comment by bennovakovic** [comment]
That seems reasonable to me. I don't believe the `refreshAccessToken` would require the cors credentials option as cookies are not required for exchanging the token, as the `refreshAccessToken` call simply passes the refreshToken to the server, and the server would reissue a new refreshToken & accessToken

**Comment by bpedroza** [comment]
Ok, that makes sense to me. So  `refreshAccessToken` would just set the `mode` option and `exchangeForAccessToken` would set the `mode` option and optionally set the `credentials` option?

**Comment by bpedroza** [comment]
Do you plan to make this updates?

**Comment by bennovakovic** [comment]
Ah sorry, I misunderstood your expectations here. I didn't realise you were looking at me to make these changes. 

I'll get to it!

**Comment by bennovakovic** [comment]
Will reopen with a clean history off `master`.

**Comment by bennovakovic** [comment]
New PR:  https://github.com/bpedroza/js-pkce/pull/42
